### PR TITLE
nixos/networkmanager: add declarative profiles

### PR DESCRIPTION
**Pull request by Janik-Haag**
## Description of changes

This PR adds a declarative way of defining network-manager profiles with secrets management. I would appreciate feedback on the nix code :)

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages

- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Janik-Haag** [comment]
cc @NixOS/freedesktop

File: nixos/modules/services/networking/networkmanager.nix
Comment (maralorn) [inline nixos/modules/services/networking/networkmanager.nix:417]: Can you maybe give a link to some documentation that describes valid fields here?

File: nixos/modules/services/networking/networkmanager.nix
Comment (maralorn) [inline nixos/modules/services/networking/networkmanager.nix:402]: There is a little name-clash here because the environment file will contain

$WORK_PASSWORD
$HOME_PASSWORD
$HOTEL1_PASSWORD

etct, right? Maybe then $PASSWORD is not an optimal example?

Comment (maralorn) [reply nixos/modules/services/networking/networkmanager.nix:402]: An alternative would be one environment file per profile. Not sure …

File: nixos/modules/services/networking/networkmanager.nix
Comment (maralorn) [inline nixos/modules/services/networking/networkmanager.nix:425]: I honestly never heard of that specific tool maybe you can also add a docs link? (Definitely only a nice to have.)

File: nixos/modules/services/networking/networkmanager.nix
Comment (maralorn) [inline nixos/modules/services/networking/networkmanager.nix:561]: In theory you could instead use the systemd LoadCreds functionality and then maybe don’t load the secrets into the environment for the whole unit but just for envsubst. From my understanding that would be current best-practice.
In practice though this unit is very lightweight (thus little code to compromise the environment) and you are writing the secrets to a root readable position anyways … So this suggestion is probably overkill.

File: nixos/modules/services/networking/networkmanager.nix
Comment (maralorn) [inline nixos/modules/services/networking/networkmanager.nix:385]: I wonder if there are some fields which are definitely mandatory for a profile to work and if we can add assertions that those are present.

File: nixos/modules/services/networking/networkmanager.nix
Comment (maralorn) [reply nixos/modules/services/networking/networkmanager.nix:417]: Also you should explain that this will only create the profiles in memory and never delete them.

Question: What happens when the user changes a profile manually after it got created with ensureProfiles? Will the profile and the change get persisted? Will the next nixos-rebuild overwrite the change? Is the profile read only? If its readonly do networkmanager clients have an acceptable UI for that?

**Review by maralorn — COMMENTED** [review]
I like this change, thanks! It doesn’t look too intrusive and I would likely use it. (Sharing wifi setups with multiple devices and household  members sounds awesome.)
I gave a few suggestions but nothing deal-breaking.

However, I haven’t tried it out and I definitely think a few people should before we merge it.
Also, a NetworkManager maintainer (if we have something like that) should look over this.

I don’t think that it is a problem that profiles don’t get removed until reboot. That’s why we call it "ensureProfiles" and there is precedent for that e.g. for cups printers.

File: nixos/modules/services/networking/networkmanager.nix
Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:417]: > Can you maybe give a link to some documentation that describes valid fields here?

I think I'll just add a link to redhats docs if that is fine? 
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_and_managing_networking/assembly_networkmanager-connection-profiles-in-keyfile-format_configuring-and-managing-networking
Other then that you can use your current profiles that are most likely in `/etc/NetworkManager/system-connections` and you can use github.com/janik-haag/nm2nix to convert those to the needed nix code. 

> Question: What happens when the user changes a profile manually after it got created with ensureProfiles? 

At first nothing (only applies for edits with editors like vim tools like nmtui follow later) because network-manager needs a reload with `nmcli connection reload` to scan them and yes then they would work but only as long as `NetworkManager-ensure-profiles` doesn't run again because it [just overrides files](https://github.com/NixOS/nixpkgs/pull/254647/files#diff-0a708e7b053cf5df7620b5262936553af2242d2ce9dabde5bbeba221ece0a021R552). The unit will run at every boot and on activation so a nixos-rebuild will also trigger it. 

To be clear if a user manually adds a profile with tools like nm-applet or the gnome one they will be stored in `/etc/NetworkManager/system-connections` and won't be touched by this tool so it is possible to use this in combination with imperative/adhoc managment and at least nm-applet just displays all profiles even when there are name collisions (I didn't test this with the gnome gui or any other). 

> If its readonly do networkmanager clients have an acceptable UI for that?

More or less, if a readonly file in temp gets edited with nmtui or nm-applet they just make the file writable and move it to the persistent directory which shouldn't bother us too much.

Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:417]: I'll add this in a structured form to the description

File: nixos/modules/services/networking/networkmanager.nix
Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:402]: > etct, right? Maybe then $PASSWORD is not an optimal example?

yeah you are right, I'll change that. 

> An alternative would be one environment file per profile. Not sure …

then we would need one unit per connection which doesn't seem ideal

File: nixos/modules/services/networking/networkmanager.nix
Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:425]: Sure, I just followed @999eagle 's example in https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/services/web-servers/traefik.nix

File: nixos/modules/services/networking/networkmanager.nix
Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:561]: I'll give this a look, never heard of LoadCreds :see_no_evil: sounds interesting, I'll add it if it doesn't add too much complexity.

File: nixos/modules/services/networking/networkmanager.nix
Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:385]: In my opinion the thing that would make the most sense to check is:
```nix
      connection = {
          id = "realraum";
          type = "wifi";
          interface-name = "wlp3s0";
        };
```
I think checking if connection exits and has the id attribute makes sense, everything else seems like a lot of work for something that's not really specified.

Adding checks for wifi, ipv4, ipv6, gsm, etc just seems overkill. There is also networkamanger plugins that add custom types and and you can just create as many interfaces adhoc as you want so checking interface-name and type also doesn't make too much sense in my opinion

**Comment by Janik-Haag** [comment]
> I like this change, thanks! It doesn’t look too intrusive and I would likely use it. (Sharing wifi setups with multiple devices and household members sounds awesome.)
I gave a few suggestions but nothing deal-breaking.

Yeah same that's why I build it, I also had the idea of creating a public repo which exposes some profiles for hackspaces and events like the chaos congress or the mrmcd and people can just add more with PRs to that repo but that's going a bit offtopic for this PR 

> However, I haven’t tried it out and I definitely think a few people should before we merge it.
Also, a NetworkManager maintainer (if we have something like that) should look over this.
I checked it and for me it works great, I have ~30 wlan connections imported two wire guard vpns, 1 Ethernet connection and even one gsm/lte profile. A few of them have special characters in there name that's why I added `escapeShellArg` to the code in some places which should catch most if not all things that could break here. 
The modules maintainers entry states the freedesktop team which is only one person and I pinged there github team in the first comment.

File: nixos/modules/services/networking/networkmanager.nix
Comment (maralorn) [reply nixos/modules/services/networking/networkmanager.nix:417]: Ah, I was not thinking about what happens if someone manually edits the /run files. I meant the nmtui edit.
It will persist the profile in another dir, okay. Does then the persistent profile take precedence over the one created in this profile or the other way around? Either way, I see the danger of very confused users because they can make changes that will have no effect. I think the documentation of this option should warn about that case.

Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:417]: > It will persist the profile in another dir,

no it will move the file. So at first there is only one profile with this name because the file was moved, there will be two as soon as the systemd unit runs again because it also creates on in the temp storage again.

> Does then the persistent profile take precedence over the one created in this profile or the other way around?

Nor it just show you two profiles with the same name.

> Either way, I see the danger of very confused users because they can make changes that will have no effect. I think the documentation of this option should warn about that case.

Yup I'll add that.

File: nixos/modules/services/networking/networkmanager.nix
Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:561]: If you mean https://freedesktop.org/software/systemd/man/systemd-creds.html with LoadCreds then it will probably make the module more complicated then it's worth.

File: nixos/modules/services/networking/networkmanager.nix
Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:385]: Do you have a suggestion on how I would implement this assertion?

Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:385]: I'm thinking of something like this:
```
      {
        assertion = (forEach(mapAttrsToList(n: v: { inherit v; }) cfg.ensure-profiles.profiles) (profile: builtins.hasAttr "connection" profile));
        message = ''
          Every NetworkManager profile needs at least a connection setting with a id attribute
        '';
      }
```

File: nixos/modules/services/networking/networkmanager.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/networkmanager.nix:383]:
```suggestion
      ensureProfiles = {
```
?

File: nixos/modules/services/networking/networkmanager.nix
Comment (RaitoBezarius) [reply nixos/modules/services/networking/networkmanager.nix:385]: I agree that the assert will be useful.

File: nixos/modules/services/networking/networkmanager.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/networkmanager.nix:418]:
```suggestion
            If you add a new ad-hoc connection via a GUI or nmtui or anything similar it should just work together with the declarativ ones.
```

File: nixos/modules/services/networking/networkmanager.nix
Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:385]: Do you have a implementation suggestion?

File: nixos/modules/services/networking/networkmanager.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/networkmanager.nix:419]:
```suggestion
            And if you edit a declarative profile, NetworkManager will move it to the persistent storage and treat it like a ad-hoc one,
```

File: nixos/modules/services/networking/networkmanager.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/networkmanager.nix:420]:
```suggestion
            but there will be two profiles as soon as the systemd unit from this option runs again which can be confusing since NetworkManager tools will start displaying two profiles with the same name and probably a bit diffrent settings depending on what you edited.
```

File: nixos/modules/services/networking/networkmanager.nix
Comment (RaitoBezarius) [reply nixos/modules/services/networking/networkmanager.nix:385]: Let's say something like this:
```
let
    isValidWiFiProfile =  ...;
    isValidConnection = connection: profile: hasAttr "id" connection && hasAttr "type" connection && (connection.type == "wifi" -> isValidWiFiProfile profile);
    isValidProfile = profile: (hasAttr "connection" profile && isValidConnection profile.connection profile)
```

And hop, we realize that we are reinventing submodules type checking. So let's go for a submodule here with freeform type options.

Then, the asserts should only be on "if this is a WiFi connection, this should have those fields minimally", etc.

File: nixos/modules/services/networking/networkmanager.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/networkmanager.nix:419]:
```suggestion
            And if you edit a declarative profile NetworkManager will move it to the persistent storage and treat it like a ad-hoc one,
```

File: nixos/modules/services/networking/networkmanager.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/networkmanager.nix:547]:
```suggestion
      description = "Ensure that NetworkManager declarative profiles are imported";
```

File: nixos/modules/services/networking/networkmanager.nix
Comment (RaitoBezarius) [inline nixos/modules/services/networking/networkmanager.nix:546]:
```suggestion
    systemd.services.NetworkManager-ensure-profiles = mkIf (cfg.profiles != { }) {
```

File: nixos/modules/services/networking/networkmanager.nix
Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix:556]: What do you mean with deletion? If it's not in the config anymore it will be gone after the next boot because they are saved in the temp directory of NM which gets cleared every boot otherwise they will just be overridden. I guess I should add this to the module description.

Comment (RaitoBezarius) [reply nixos/modules/services/networking/networkmanager.nix:556]: Yes, exactly, this is what I meant and I think it's good to mention it :-)

File: nixos/modules/services/networking/networkmanager.nix
Comment (RaitoBezarius) [reply nixos/modules/services/networking/networkmanager.nix]: just massage it pre serialization

Comment (RaitoBezarius) [reply nixos/modules/services/networking/networkmanager.nix]: look over all children and if you find that all subvalues are nulls, skip the section

Comment (Janik-Haag) [reply nixos/modules/services/networking/networkmanager.nix]: this should do the trick https://github.com/NixOS/nixpkgs/pull/254647/files#diff-0a708e7b053cf5df7620b5262936553af2242d2ce9dabde5bbeba221ece0a021R402 of filtering out empty attrsets now I have to figure out how to do the wifi submodule thing so it doesn't do the checks if it defaults to empty attrSet

**Review by NetaliDev — APPROVED** [review]
Works :)

**Review by RaitoBezarius — APPROVED** [review]

**Pull request merged**
